File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/mouse/TapDetector.java
package net.kdt.pojavlaunch.customcontrols.mouse;

import static android.view.MotionEvent.ACTION_DOWN;
import static android.view.MotionEvent.ACTION_POINTER_DOWN;
import static android.view.MotionEvent.ACTION_POINTER_UP;
import static android.view.MotionEvent.ACTION_UP;

import android.view.MotionEvent;

import net.kdt.pojavlaunch.Tools;

/**
 * ポインターに関係なくXタップイベントをより正確に検出するためのクラス
 * 可能な限り最小限のイベントのみを使用します。
 * すべてのイベントが順番通りに届くとは限らないため
 */
public class TapDetector {

    public final static int DETECTION_METHOD_DOWN = 0x1;
    public final static int DETECTION_METHOD_UP = 0x2;
    public final static int DETECTION_METHOD_BOTH = 0x3; //現在未使用

    private final static int TAP_MIN_DELTA_MS = -1;
    private final static int TAP_MAX_DELTA_MS = 300;
    private final static int TAP_SLOP_SQUARE_PX = (int) Tools.dpToPx(2500);

    private final int mTapNumberToDetect;
    private int mCurrentTapNumber = 0;

    private final int mDetectionMethod;

    private long mLastEventTime = 0;
    private float mLastX = 9999;
    private float mLastY = 9999;

    /**
     * @param tapNumberToDetect onTouchEventがTrueを返す前に必要なタップ数
     * @param detectionMethod タッチ検出に使用する方法。上記のDETECTION_METHOD定数を参照。
     */
    public TapDetector(int tapNumberToDetect, int detectionMethod){
        this.mDetectionMethod = detectionMethod;
        //DETECTION_METHOD_BOTHではACTION_DOWNとACTION_UPの両方を期待
        this.mTapNumberToDetect = detectBothTouch() ? 2*tapNumberToDetect : tapNumberToDetect;
    }

    /**
     * タッチイベント発生時に呼び出す関数です。
     * @param e 検査するMotionEvent
     * @return ポインターでXタップが発生したかどうか
     */
    public boolean onTouchEvent(MotionEvent e){
        int eventAction = e.getActionMasked();
        int pointerIndex = -1;

        //前方参照するイベントを取得
        if(detectDownTouch()){
            if(eventAction == ACTION_DOWN) pointerIndex = 0;
            else if(eventAction == ACTION_POINTER_DOWN) pointerIndex = e.getActionIndex();
        }
        if(detectUpTouch()){
            if(eventAction == ACTION_UP) pointerIndex = 0;
            else if(eventAction == ACTION_POINTER_UP) pointerIndex = e.getActionIndex();
        }

        if(pointerIndex == -1) return false; // Useless event

        // 現在のイベント情報を保存
        float eventX = e.getX(pointerIndex);
        float eventY = e.getY(pointerIndex);
        long eventTime = e.getEventTime();

        // デルタ値を計算
        long deltaTime = eventTime - mLastEventTime;
        int deltaX = (int) mLastX - (int) eventX;
        int deltaY = (int) mLastY - (int) eventY;

        // 現在のイベント情報を保存 to persist on next event
        mLastEventTime = eventTime;
        mLastX = eventX;
        mLastY = eventY;

        // 十分な速度と精度があるか確認
        if(mCurrentTapNumber > 0){
            if  ((deltaTime < TAP_MIN_DELTA_MS || deltaTime > TAP_MAX_DELTA_MS) ||
                ((deltaX*deltaX + deltaY*deltaY) > TAP_SLOP_SQUARE_PX)) {
                if (mDetectionMethod == DETECTION_METHOD_BOTH && (eventAction == ACTION_UP || eventAction == ACTION_POINTER_UP)) {
                    // For the both method, the user is expected to start with a down action.
                    resetTapDetectionState();
                    return false;
                } else {
                    // We invalidate previous taps, not this one though
                    mCurrentTapNumber = 0;
                }
            }
        }

        // 有意義なタップが発生
        mCurrentTapNumber += 1;
        if(mCurrentTapNumber >= mTapNumberToDetect){
           resetTapDetectionState();
           return true;
        }

        // 十分なタップ数に達していない場合
        return false;
    }

    /**
     * ダブルタップの値をリセットします。
     */
   private void resetTapDetectionState(){
       mCurrentTapNumber = 0;
       mLastEventTime = 0;
       mLastX = 9999;
       mLastY = 9999;
   }
/**
 * 「detect Down Touch」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
   private boolean detectDownTouch(){
       return (mDetectionMethod & DETECTION_METHOD_DOWN) == DETECTION_METHOD_DOWN;
   }
/**
 * 「detect Up Touch」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
   private boolean detectUpTouch(){
       return (mDetectionMethod & DETECTION_METHOD_UP) == DETECTION_METHOD_UP;
   }
/**
 * 「detect Both Touch」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
   private boolean detectBothTouch(){
       return mDetectionMethod == DETECTION_METHOD_BOTH;
   }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/mouse/TapDetector.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.mouse

import android.view.MotionEvent.ACTION_DOWN
import android.view.MotionEvent.ACTION_POINTER_DOWN
import android.view.MotionEvent.ACTION_POINTER_UP
import android.view.MotionEvent.ACTION_UP
import android.view.MotionEvent
import com.arata.yukarilauncher.Tools

/** タップ操作を検出するクラス。シングルタップ、ダブルタップを時間・距離で判定します。 */
class TapDetector(tapNumberToDetect: Int, private val mDetectionMethod: Int) {
    companion object {
        /** ダウンイベントで検出。 */
        const val DETECTION_METHOD_DOWN = 0x1
        /** アップイベントで検出。 */
        const val DETECTION_METHOD_UP = 0x2
        /** ダウンとアップの両方で検出。 */
        const val DETECTION_METHOD_BOTH = 0x3
        private const val TAP_MIN_DELTA_MS = -1
        private const val TAP_MAX_DELTA_MS = 300
        private val TAP_SLOP_SQUARE_PX = Tools.dpToPx(2500f).toInt()
    }

    private val mTapNumberToDetect: Int
    private var mCurrentTapNumber = 0
    private var mLastEventTime: Long = 0
    private var mLastX = 9999f
    private var mLastY = 9999f

    init {
        mTapNumberToDetect = if (detectBothTouch()) 2 * tapNumberToDetect else tapNumberToDetect
    }

    /**
     * タッチイベントを処理してタップを検出します。
     * @param e 処理するタッチイベント
     * @return タップが検出された場合はtrue
     */
    fun onTouchEvent(e: MotionEvent): Boolean {
        val eventAction = e.actionMasked
        var pointerIndex = -1

        if (detectDownTouch()) {
            if (eventAction == ACTION_DOWN) pointerIndex = 0
            else if (eventAction == ACTION_POINTER_DOWN) pointerIndex = e.actionIndex
        }
        if (detectUpTouch()) {
            if (eventAction == ACTION_UP) pointerIndex = 0
            else if (eventAction == ACTION_POINTER_UP) pointerIndex = e.actionIndex
        }

        if (pointerIndex == -1) return false

        val eventX = e.getX(pointerIndex)
        val eventY = e.getY(pointerIndex)
        val eventTime = e.eventTime

        val deltaTime = eventTime - mLastEventTime
        val deltaX = (mLastX - eventX).toInt()
        val deltaY = (mLastY - eventY).toInt()

        mLastEventTime = eventTime
        mLastX = eventX
        mLastY = eventY

        if (mCurrentTapNumber > 0) {
            if ((deltaTime < TAP_MIN_DELTA_MS || deltaTime > TAP_MAX_DELTA_MS) ||
                (deltaX * deltaX + deltaY * deltaY > TAP_SLOP_SQUARE_PX)) {
                if (mDetectionMethod == DETECTION_METHOD_BOTH && (eventAction == ACTION_UP || eventAction == ACTION_POINTER_UP)) {
                    resetTapDetectionState()
                    return false
                } else {
                    mCurrentTapNumber = 0
                }
            }
        }

        mCurrentTapNumber += 1
        if (mCurrentTapNumber >= mTapNumberToDetect) {
            resetTapDetectionState()
            return true
        }

        return false
    }

    /** タップ検出状態をリセットします。 */
    private fun resetTapDetectionState() {
        mCurrentTapNumber = 0
        mLastEventTime = 0
        mLastX = 9999f
        mLastY = 9999f
    }

    /** @return ダウンイベントを検出するかどうか */
    private fun detectDownTouch(): Boolean {
        return (mDetectionMethod and DETECTION_METHOD_DOWN) == DETECTION_METHOD_DOWN
    }

    /** @return アップイベントを検出するかどうか */
    private fun detectUpTouch(): Boolean {
        return (mDetectionMethod and DETECTION_METHOD_UP) == DETECTION_METHOD_UP
    }

    /** @return 両方のイベントを検出するかどうか */
    private fun detectBothTouch(): Boolean {
        return mDetectionMethod == DETECTION_METHOD_BOTH
    }
}